scala/private/phases/api.bzl

```diff
@@ -0,0 +1,47 @@
+load(
+    "@io_bazel_rules_scala//scala:providers.bzl",
+    _ScalaRulePhase = "ScalaRulePhase",
+)
+
+def _adjust_phases(phases, adjustments):
+    if len(adjustments) == 0:
+        return phases
+    phases = phases[:]
+    for (relation, peer_name, name, function) in adjustments:
+        for idx, (needle, _) in enumerate(phases):
+            if needle == peer_name:
+                if relation in ["-", "before"]:
```

Comment:
The phase extension mechanism is perhaps a bit overkill. It does allow some neat exploration by wrapping up scalac plugins with providers that augment the phases. This way, for example, the semanticdb plugin could intelligently augment the rules and save the semantic database file as an additional output.

This same functionality can also be achieved by simply building your own rule a la carte using the phases directly.

The TLDR is that I like the phase extension mechanism, but I don't know if it has fully proven its worth. If it gets merged as part of this PR I propose we evaluate keeping/removing it at a later date once we have more data.

Reply:
My original comments were mainly about the operators (symbols in addition to keywords) but I think my more general concern is that we are making rules_scala API dramatically larger by adding all phases, extension mechanism (both via keywords and via symbols).
I'm concerned about maintenance costs.
Current PR also makes it really easy for someone to customize in places where maybe they shouldn't. I'm trying to think how to make customization possible but less accessible so that we don't have many splinters which mean we can't progress.
I'm not sure exactly what this means but one example can be to not export all the phases everywhere, clearly separate phases (files as well as docs) from "regular" API (maybe this is the way it is now I'm not sure since the PR is huge so I'm a bit lost in the details) and another
example is that I'd consider dropping the symbols (not my biggest concern but an example).
wdyt?

Reply:
> are all of these operators covered in tests?
are they documented for user extensibility?

Not yet, but I am happy to do so.

> I'd consider dropping the symbols

I am fine with it, but I don't think it matters too much.

> easy for someone to customize in places where maybe they shouldn't

People may easily break things if they don't fully understand how things work, but I don't think it should prevent us from making it really flexible for those who have the knowledge. We still provide the official default rules, and that's the part we should always make sure it's awesome.

Phases extension mechanism also makes maintainers/developers easy to make changes. Check out the example for formatting https://github.com/bazelbuild/rules_scala/pull/874.

Reply:
@borkaehw I think I'm having a hard time getting my point across.
I'm in favor of the phases and even the extension mechanism seems valuable for exactly the reasons you mention.
What I'm concerned about is making our API surface much bigger.
Maybe there's a way we can signal "danger zone" for most users (who obviously don't need this level of extensions).
Does that make sense?

Reply:
Sure, I understand your concern

> making our API surface much bigger

IMHO stating it clear in the documents and READMEs might be sufficient to prevent bad experience from general users. I could be wrong.

> we can signal "danger zone" for most users

I am interested in how it will look like. :smile:

Reply:
One suggestion is to have all extension related under `//scala/advanced-usage`. This is to signal it's not part of the standard d2d usage for most companies.
Also, are the symbols here based on some kind of convention (npm or something) or something we made up?

Reply:
> //scala/advanced-usage

I like it. I will update the PR.

> are the symbols here based on some kind of convention

I think @andyscott would be a better person for this question.